lib/ansible/plugins/inventory/foreman.py

```diff
@@ -121,134 +126,139 @@ def _get_session(self):
+        s = self._get_session()
+
+        while True:
+            ret = s.get("%s/api/v2/hosts" % self.foreman_url, params=params)
+            ret.raise_for_status()
+            json = ret.json()
+
+            # process results
+            # FIXME: This assumes 'return type' matches a specific query,
+            #        it will break if we expand the queries and they dont have different types
+            if 'results' not in json:
+                # /hosts/:id dos not have a 'results' key
+                results = json
+                break
+            elif isinstance(json['results'], MutableMapping):
+                # /facts are returned as dict in 'results'
+                results = json['results']
+                break
+            else:
+                # /hosts 's 'results' is a list of all hosts, returned is paginated
+                results = results + json['results']
+
+                # check for end of paging
+                if len(results) >= json['subtotal']:
                     break
-                elif isinstance(json['results'], MutableMapping):
-                    # /facts are returned as dict in 'results'
-                    results = json['results']
+                if len(json['results']) == 0:
+                    self.display.warning("Did not make any progress during loop. expected %d got %d" % (json['subtotal'], len(results)))
                     break
-                else:
-                    # /hosts 's 'results' is a list of all hosts, returned is paginated
-                    results = results + json['results']
 
-                    # check for end of paging
-                    if len(results) >= json['subtotal']:
-                        break
-                    if len(json['results']) == 0:
-                        self.display.warning("Did not make any progress during loop. expected %d got %d" % (json['subtotal'], len(results)))
-                        break
-
-                    # get next page
-                    params['page'] += 1
+                # get next page
+                params['page'] += 1
 
+        if self.use_cache:
             self._cache[self.cache_key][url] = results
 
-        return self._cache[self.cache_key][url]
+        return results
+
+    def to_safe(self, word):
+        '''Converts 'bad' characters in a string to underscores so they can be used as Ansible groups
+        #> ForemanInventory.to_safe("foo-bar baz")
+        'foo_barbaz'
+        '''
+        regex = r"[^A-Za-z0-9\_]"
+        return re.sub(regex, "_", word.replace(" ", ""))
+
+    def _resolve_params(self, host):
+        '''
+        Fetch host params and convert to dict, then add to inventory
+        '''
+        for param in host['all_parameters']:
+            key = param['name']
+
+            # Try json
+            try:
+                value = json.loads(param['value'])
+            except ValueError:
+                # Not JSON, try YAML
+                pass
+            try:
```

Comment:
This should be in the `except` above.

Reply:
I don't know if I understand what you mean here. Do you mean that "ValueError" should be used in line 134 ?